Add whylogs SessionManager

murilommen opened the pull request:
## Description

A `SessionManager` to make it easier to manage credentials to authenticate with WhyLabs

## Changes

- Adding a `SessionManager` as a singleton for the `init` method
- Added `UserSession` and `GuestSession` as containers for each auth flow, as well as helper methods to them

- [x] I have reviewed the [Guidelines for Contributing](CONTRIBUTING.md) and the [Code of Conduct](CODE_OF_CONDUCT.md).

murilommen committed:
add Singletion Session object for login

murilommen committed:
refactor: Session -> SessionManager

naddeoa commented on python/whylogs/api/whylabs/session_manager.py:
```diff
@@ -0,0 +1,113 @@
+from .variables import Variables
+
+_auth_path = get_auth_file_path(auth_path=Path(f"{Path.home()}/.whylabs/auth.ini"))
+DEFAULT_WHYLABS_HOST = "https://songbird.development.whylabsdev.com"
```
Just a temporary dev thing?

naddeoa commented on python/whylogs/api/whylabs/session_manager.py:
```diff
@@ -0,0 +1,113 @@
+@dataclass
+class UserSession:
+    org_id: str
+    dataset_id: str
```
Will this be a thing? I kind of assumed that dataset would continue to live in the api latyer rather than have a single value since it's far more likely to change than the org/api key

naddeoa commented on python/whylogs/api/whylabs/session_manager.py:
```diff
@@ -0,0 +1,113 @@
+    def __post_init__(self):
+        Variables.set_variable_to_config_file(key="org_id", value=self.org_id, auth_path=_auth_path)
+        Variables.set_variable_to_config_file(key="dataset_id", value=self.dataset_id, auth_path=_auth_path)
+        keyring.set_password("whylabs", "api_key", self.api_key)
```
Does this end up prompting them for a password when we call `get`? And is this the right spot to put some of the `if_notebook` checks? I assume we'll have to do that for some of the UX for entering settings using `input()`.

naddeoa commented on python/whylogs/api/whylabs/session_manager.py:
```diff
@@ -0,0 +1,113 @@
+    api = SessionsApi(client)
+    response: CreateSessionResponse = api.create_session(CreateSessionRequest(user_id))
+
+    print(f"Created session {response.id}")
```
Do we use `logging` for this stuff?

murilommen replied:
better to! will change that

naddeoa commented on python/whylogs/api/whylabs/session_manager.py:
```diff
@@ -0,0 +1,113 @@
+
+def create_session(anonymous: Optional[str] = None) -> Union[GuestSession, UserSession]:
+    if anonymous is None:
+        anonymous = input("Do you want to create an anonymous session? [True/False] (default: False) ")
```
ah this. But I think we probably want gate it in that `if_notebook` check in case they run this in a normal file after. Not even sure what would happen.

naddeoa replied:
Especially if we have a while True loop lol

naddeoa commented on python/whylogs/api/whylabs/session_manager.py:
```diff
@@ -0,0 +1,113 @@
+    @staticmethod
+    def get_instance(anonymous: Optional[str] = None):
+        if SessionManager.__instance is None:
+            SessionManager(anonymous=anonymous)
```
Doesn't this have to be set to `Sessionmanager.__instance`?

murilommen replied:
is is set on `SessionManager`'s init

naddeoa reviewed (commented):
Sanity check before we send all of this code. Is this basically doing what pydantic's BaseSettings class does automatically?

jamie256 commented on python/whylogs/api/whylabs/session_manager.py:
```diff
@@ -0,0 +1,113 @@
+from pathlib import Path
+from typing import Optional, Union
+
+import keyring
```
for new run time dependencies (such as keyring) please do the following:
1. `poetry add keyring`
2. `poetry update`

Also test in a non-dev environment, e.g. after you build a wheel off your branch you can drag it onto colab as a local file, then install that wheel and try to use it, minimally run:
`import whylogs`

The problem with this change is that this dependency is already a dev dependency (pulled in by twine), so things will work locally for whylogs developers poetry env, but keyring is missing in the pyproject.toml file's install time depedencies, so new builds would break whylogs import if the env doesn't already have the required packages.

murilommen replied:
right! i'm not 100% confident in adding keyring just yet. will test it on other environments (ubuntu, wsl, windows) shortly

murilommen committed:
is_notebook to fork variable fetching

murilommen committed:
Merge remote-tracking branch 'origin/mainline' into dev/murilommen/whylogs-config

# Conflicts:
#	python/poetry.lock

murilommen committed:
adding unit tests

murilommen committed:
Merge branch 'mainline' into dev/murilommen/whylogs-config

murilommen committed:
Merge branch 'mainline' into dev/murilommen/whylogs-config

murilommen committed:
Merge branch 'mainline' into dev/murilommen/whylogs-config

murilommen committed:
Merge branch 'mainline' into dev/murilommen/whylogs-config

murilommen committed:
incresing timeout for notebook tests

murilommen committed:
Revert "incresing timeout for notebook tests"

This reverts commit fcd9623efce501fd15be0d931fe941aa1b73da0d.

murilommen committed:
attempt: check if init changes nb-test

murilommen committed:
removing while loop:

* infinite while loop might crash notebooks or cause weird things
currently it's only there to prompt if anonymous

murilommen committed:
logging things to console on init

murilommen committed:
removing input section

murilommen committed:
adding a github actions check

murilommen committed:
adding interactive flag

murilommen committed:
removing while loop

murilommen committed:
Merge branch 'mainline' into dev/murilommen/whylogs-config

murilommen committed:
fix version on fugue profiling

murilommen committed:
commenting out ls command

andyndang commented on python/tests/api/whylabs/test_session_manager.py:
```diff
@@ -0,0 +1,109 @@
+    def setup_class(cls):
+        os.environ["WHYLABS_API_KEY"] = "api_key"
+        os.environ["ORG_ID"] = "org_id"
+        os.environ["WHYLABS_API_ENDPOINT"] = "https://songbird.development.whylabsdev.com"
```
should leave our dev endpoint here

andyndang commented on python/whylogs/api/whylabs/session_manager.py:
```diff
@@ -0,0 +1,115 @@
+    config = Configuration()
+    config.host = os.getenv("WHYLABS_HOST") or DEFAULT_WHYLABS_HOST
+
+    client = ApiClient(config)
```
don't create the API client every time. It's expensive - should reuse the connection pool in general

murilommen replied:
this is only created and called once - when users want to create an anonymous session AND they don't have it on the config file yet

andyndang commented on python/whylogs/api/whylabs/session_manager.py:
```diff
@@ -0,0 +1,115 @@
+    return GuestSession(session_id=session_id)
+
+
+def create_session(anonymous: Optional[bool] = None, interactive: bool = False) -> Union[GuestSession, UserSession]:
```
would this fail gracefully if they can't connect to our endpoint?

andyndang commented on python/whylogs/api/whylabs/session_manager.py:
```diff
@@ -0,0 +1,115 @@
+
+
+def init(anonymous: Optional[bool] = None, interactive: bool = False) -> None:
+    SessionManager.get_instance(anonymous=anonymous, interactive=interactive)
```
Do we always create new sessions if `anonymous` flag is enabled? We're not pulling config from the file regardless?

murilommen replied:
no, if the anonymous flag is enabled, it will first try to check if there is a session_id on the default path, otherwise it will create it using the API

andyndang commented on python/whylogs/api/whylabs/session_manager.py:
```diff
@@ -0,0 +1,115 @@
+    if not anonymous:
+        return _get_logged_session(interactive=interactive)
+    else:
+        return _get_guest_session()
```
We should enable re-using anonymous sessions

murilommen replied:
but that's what the singleton `SessionManager` is used for. if there is an active session (either logged or guest), it will refer to it instead of creating a new one

andyndang replied:
I don't think that's the logic here. Looks like it only re-using the session in the same process. but if the process terminates then the session is forever gone

murilommen replied:
that's true. but when starting the process over, the first thing will be checking the config file for an existing `session_id`. let's keep it on the same thread, but please let me know if this should be handled differently

andyndang commented on python/whylogs/api/whylabs/variables.py:
```diff
@@ -0,0 +1,26 @@
+class Variables:
+    @staticmethod
+    def get_variable_from_input(variable_name: str) -> str:
+        variable = input(f"What is your {variable_name}? ")
```
This isn't good. For API key you should use `getPass` or it'll get saved in the notebook https://docs.python.org/3/library/getpass.html

jamie256 commented on python/whylogs/api/whylabs/session_manager.py:
```diff
@@ -0,0 +1,115 @@
+DEFAULT_PATH = os.getenv("WHYLOGS_CONFIG_PATH") or f"{Path.home()}/.whylabs/auth.ini"
+_auth_path = get_auth_file_path(auth_path=Path(DEFAULT_PATH))
+
+DEFAULT_WHYLABS_HOST = "https://api.whylabsapp.com"
```
consider:
```suggestion
DEFAULT_WHYLABS_HOST = os.environ.get("WHYLABS_API_ENDPOINT") or "https://api.whylabsapp.com"
```

Consider moding this into a common file so we can use same definition for session manager and the whylabs writer.

murilommen replied:
ok! makes sense

jamie256 commented on python/examples/integrations/Fugue_Profiling.ipynb:
```diff
@@ -1071,7 +1071,7 @@
    ],
    "source": [
-    "!ls /tmp/output*.parquet"
+    "# !ls /tmp/output*.parquet"
```
?

murilommen replied:
this is the command that made my changes fail. i have no clue why and I couldn't repro this locally. on my branch the tests pass. 
apparently it's something that only happens on Ubuntu environments? since Anthony could narrow it down to this line of code. 
i feel awkward to make this change, but can't find a way to make this work without it.

jamie256 reviewed (commented):
Why do we need to modify the fugue example notebook here

murilommen committed:
Merge branch 'mainline' into dev/murilommen/whylogs-config

murilommen committed:
host variable, getpass

details:
- WHYLABS_HOST -> WHYLABS_API_ENDPOINT to make it standard with
current whylogs usage on WhyLabsWriter

- using getpass.getpass() for the api_key input flow

murilommen committed:
Merge remote-tracking branch 'origin/dev/murilommen/whylogs-config' into dev/murilommen/whylogs-config

murilommen committed:
making create_session_id fail gracefully

murilommen committed:
removing dev endpoint call

andyndang commented on python/whylogs/api/whylabs/session_manager.py:
```diff
@@ -0,0 +1,120 @@
+            raise Exception("There is an active Session, use Session.get_instance() instead")
+        else:
+            SessionManager.__instance = self
+            self.session = create_session(anonymous=self._anonymous, interactive=self._interactive)
```
so you are always creating a new session if anonymous is true? But that means users will lose their session context if they terminate the notebook. We should store the session ID in the config and let user re-use them in the future in this case rather than creating new session every time

murilommen replied:
but that's what `create_session` does for a guest session. on line 79:
```python
def _get_guest_session() -> GuestSession:
    session_id = Variables.get_variable_from_config_file(auth_path=_auth_path, key="session_id")
    if session_id is None:
        (...)
```

if you have any other suggestions on how i should do it, let me know

murilommen replied:
I've just renamed the method for clarification on this behavior

murilommen committed:
renaming create_session for clarity

details:
- create_session was actually trying to get a session or create it
especially for anonymous=True, when
it tries to read from the config file before making the
API request

andyndang commented on python/whylogs/api/whylabs/session_manager.py:
```diff
@@ -0,0 +1,122 @@
+    anonymous: Optional[bool] = None, interactive: bool = False
+) -> Union[GuestSession, UserSession]:
+    if not anonymous:
+        return _get_logged_session(interactive=interactive)
```
`get_or_create`?

murilommen replied:
in the case of a logged session it will always be "get", since it basically tries to get it from either an env var, config file or input (if in interactive mode)

andyndang commented on python/whylogs/api/whylabs/session_manager.py:
```diff
@@ -0,0 +1,122 @@
+class GuestSession:
+    def __init__(self, session_id: str):
+        self.session_id = session_id
+        Variables.set_variable_to_config_file(key="session_id", value=self.session_id, auth_path=_auth_path)
```
The flow for this saving logic is really strange here. You should save it only at user input prompt, not when creating an object like this. By doing this at the object init level you are generating side effects - what if somehow we need to create these sessions manually for whatever reason? But we don't want to overwrite the existing configs?

murilommen replied:
ok, I'll move it out of the init instead!

andyndang commented on python/whylogs/api/whylabs/session_manager.py:
```diff
@@ -0,0 +1,122 @@
+
+    if is_notebook() and interactive is True:
+        api_key = api_key or Variables.get_variable_from_getpass(variable_name="api_key")
+        org_id = org_id or Variables.get_variable_from_input(variable_name="org_id")
```
You should only save the variables at this point

andyndang commented on python/whylogs/api/whylabs/session_manager.py:
```diff
@@ -0,0 +1,122 @@
+    if session_id is None:
+        user_id = str(uuid.uuid4())
+        session_id = _create_session_id(user_id=user_id)
+    return GuestSession(session_id=session_id)
```
or at this point (see the comment above about not creating side effects in a constructor

andyndang commented on python/whylogs/api/whylabs/session_manager.py:
```diff
@@ -0,0 +1,122 @@
+        return response.id
+    except ApiException as e:
+        logger.error(e)
+        raise e
```
This is bad. Nothing is catching the error above. 

We should always fall back to local mode if this call cannot be made and just log out a warning

murilommen replied:
What does local mode mean for the anonymous login? this API is only called when creating a GuestSession.
also, if not with ApiException, what else could I use to catch errors here? I'm using the same pattern indicated on other songbird calls across our examples

andyndang replied:
Got it. I thought this was getting called everytime

murilommen committed:
Merge branch 'mainline' into dev/murilommen/whylogs-config

murilommen committed:
setting variables out of Session's init

murilommen committed:
removing comment from fugue example

andyndang commented on python/whylogs/api/whylabs/session_manager.py:
```diff
@@ -0,0 +1,119 @@
+
+
+@dataclass
+class UserSession:
```
Would rename this to `ApiKeySession` to make it explicit since there are other ways a user can auth with a service

andyndang commented on python/whylogs/api/whylabs/session_manager.py:
```diff
@@ -0,0 +1,119 @@
+        if SessionManager.__instance is not None:
+            raise Exception("There is an active Session, use Session.get_instance() instead")
+        else:
+            SessionManager.__instance = self
```
this should be after `get_or_create_session` because that method can throw exception

andyndang commented on python/whylogs/api/whylabs/auth_file.py:
```diff
@@ -0,0 +1,27 @@
+
+def get_configparser_object(auth_path: Path) -> configparser.ConfigParser:
+    config = configparser.ConfigParser()
+    config.read(auth_path)
```
Does this throw exception if the file is corrupted or missing?

andyndang commented on python/whylogs/api/whylabs/auth_file.py:
```diff
@@ -0,0 +1,27 @@
+from pathlib import Path
+
+
+def _create_blank_auth_file(auth_path: Path) -> Path:
```
what happens if this method fails due to file system permission? No home folder?

andyndang commented on python/whylogs/api/whylabs/session_manager.py:
```diff
@@ -0,0 +1,119 @@
+from whylogs.api.whylabs.variables import Variables
+
+DEFAULT_PATH = os.getenv("WHYLOGS_CONFIG_PATH") or f"{Path.home()}/.whylabs/auth.ini"
+_auth_path = get_auth_file_path(auth_path=Path(DEFAULT_PATH))
```
This is not a good practice to run code with side effect at import. 

This will break whylogs if `get_auth_file_path` fails. There are different failure modes:
- Permission issues
- The file OS could hang and block the main application

You should move this logic to JIT - only runs this check when user calls `why.init` (or whenever the `init` flow runs)

andyndang reviewed (changes requested):
More comment around init/error handliing

murilommen committed:
make auth_path JIT with init execution

murilommen committed:
more guardrails for filesystem errors

details:
- now create_blank_auth_file and config_parse getter will
both throw if there is either one of:
configparser.Error, OSError, IOError, PermissionError

andyndang commented on python/whylogs/api/whylabs/auth_file.py:
```diff
@@ -0,0 +1,39 @@
+        config.read(auth_path)
+        return config
+    except (configparser.Error, OSError, IOError, PermissionError) as e:
+        raise ValueError(f"Error reading config file from {auth_path}: {str(e)}") from e
```
Shouldn't you log a warning and return an empty object instead?

andyndang reviewed (changes requested):
Just want to make sure we are on the same page here - I want to make sure that when user does:

```
why.init(...)
```

they will not be BLOCKED from using whylogs because of initialization errors due to invalid configs/network errors/file system errors. They might not be able to write to whylabs or use anonymous session, but they should never be blocked from using the main functionalities

murilommen committed:
make env vars avoid dealing with local files

andyndang commented on python/whylogs/api/whylabs/session_manager.py:
```diff
@@ -0,0 +1,126 @@
+
+        if is_notebook():
+            if api_key is None:
+                api_key = Variables.get_variable_from_getpass(variable_name="api_key")
```
you should capture these strings as constrants. It's really easy to have a typo

andyndang reviewed (commented):
Overall it looks better now but I'll leave the main approval to Jamie

murilommen committed:
Merge branch 'mainline' into dev/murilommen/whylogs-config
